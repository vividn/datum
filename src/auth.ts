import pass from "./utils/pass";
import Nano, { DocumentScope } from "nano";
import { EitherPayload } from "./documentControl/DatumDocument";
import dotenv from "dotenv";
import { BaseDatumArgs } from "./input";

export const connectDb = async (
  args: BaseDatumArgs
): Promise<DocumentScope<EitherPayload>> => {
  if (args.env !== undefined) {
    dotenv.config({ path: args.env });
  }
  const env = process.env.NODE_ENV || "production";
  const defaultHost =
    env === "production" ? "localhost:5984" : "localhost:5983";
  const couchConfig = {
    username: args.username ?? process.env.COUCHDB_USER ?? "admin",
    password: args.password ?? process.env.COUCHDB_PASSWORD ?? "password",
    hostname: args.host ?? process.env.COUCHDB_HOSTNAME ?? defaultHost,
  };
  const nano = Nano(
    `http://${couchConfig.username}:${couchConfig.password}@${couchConfig.hostname}`
  );

  const { db: dbName = "datum" } = args;

  if (args.autoCreateDb) {
    // Create database if it doesn't exist
    // TODO: Remove this
    await nano.db.create(dbName).catch(pass);
  }

  const db: DocumentScope<EitherPayload> = nano.use(dbName);
  return db;
};
