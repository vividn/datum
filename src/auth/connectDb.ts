import { EitherPayload } from "../documentControl/DatumDocument";
import PouchDb from "pouchdb";
import memoryAdapter from "pouchdb-adapter-memory";
import { MainDatumArgs } from "../input/mainArgs";
import { mergeConfigAndEnvIntoArgs } from "../config/mergeConfigIntoArgs";
import fs from "fs";

PouchDb.plugin(memoryAdapter);

export function connectDb(
  args: MainDatumArgs,
): PouchDB.Database<EitherPayload> {
  mergeConfigAndEnvIntoArgs(args);

  const host = args.host;
  const adapter =
    process.env["POUCHDB_ADAPTER"] || host === "%MEMORY%"
      ? "memory"
      : undefined;
  const { db: dbName = "datum", createDb } = args;

  if (host === undefined) {
    // TODO: when going into the browser, maybe support this?
    throw new Error("No hostname provided for database connection");
  }

  const fullDatabaseName =
    adapter === "memory"
      ? dbName
      : !host
        ? dbName
        : host.at(-1) === "/"
          ? `${host}${dbName}`
          : `${host}/${dbName}`;

  if (host.startsWith("/") && adapter !== "memory") {
    // create parent directories
    if (!fs.existsSync(fullDatabaseName)) {
      fs.mkdirSync(fullDatabaseName, { recursive: true });
    }
  }

  const couchAuth = {
    username: args.user,
    password: args.password,
  };
  return new PouchDb(fullDatabaseName, {
    skip_setup: !createDb,
    auth: couchAuth,
    adapter,
  });
}
