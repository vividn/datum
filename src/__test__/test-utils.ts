import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
