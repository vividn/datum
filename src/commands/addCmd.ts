import { EitherDocument } from "../documentControl/DatumDocument";
import { connectDb } from "../auth/connectDb";
import {
  addDoc,
  conflictChoices,
  ConflictStrategyNames,
} from "../documentControl/addDoc";
import { dataArgs, DataArgs, handleDataArgs } from "../input/dataArgs";
import { MainDatumArgs } from "../input/mainArgs";
import { addIdAndMetadata } from "../meta/addIdAndMetadata";
import { primitiveUndo } from "../undo/primitiveUndo";
import { FieldArgs, fieldArgs } from "../input/fieldArgs";
import { flexiblePositional } from "../input/flexiblePositional";
import { updateLastDocsRef } from "../documentControl/lastDocs";
import { compileState } from "../state/compileState";
import { ArgumentParser } from "argparse";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const newDocArgs = new ArgumentParser({
  add_help: false,
});
newDocArgs.add_argument("--no-metadata", "-M", {
  help: "do not include meta data in document",
  action: "store_true",
  dest: "noMetadata",
});
newDocArgs.add_argument("--id-part", "--id", {
  help:
    "Which field(s) to use for the _id field in the document." +
    " Can either be a single string with fields delimited by --id-delimiter" +
    " or can be used multiple times to progressively assemble an id delimited by --id-delimiter",
  action: "append",
  dest: "idParts",
});
newDocArgs.add_argument("--id-delimiter", {
  help: "spacer between fields in the id",
});
newDocArgs.add_argument("--undo", "-u", {
  help: "undoes the last datum entry",
  action: "store_true",
});
newDocArgs.add_argument("--force-undo", "-U", {
  help: "forces an undo, even if the datapoint was entered more than 15 minutes ago",
  action: "store_true",
  dest: "forceUndo",
});
newDocArgs.add_argument("--merge", "-x", {
  help: "on conflict with an existing document update with the merge strategy. Equivalent to `--update merge`",
  action: "store_const",
  const: "merge",
  dest: "conflict",
});
newDocArgs.add_argument("--conflict", "-X", {
  help: `on conflict, update with given strategy.`,
  choices: conflictChoices,
});

export const addArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, dataArgs],
});
export const addCmdArgs = new ArgumentParser({
  description: "add a document",
  prog: "datum add",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <reqValue> [optValue] [data..]
`,
  parents: [addArgs, outputArgs, dbArgs],
});

export type AddCmdArgs = MainDatumArgs &
  FieldArgs &
  DataArgs & {
    noMetadata?: boolean;
    idParts?: string[];
    idDelimiter?: string;
    undo?: boolean;
    forceUndo?: boolean;
    merge?: boolean;
    conflict?: ConflictStrategyNames;
  };

export async function addCmd(
  args: AddCmdArgs | string | string[],
  preparsed?: Partial<AddCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(addCmdArgs, args, preparsed);
  const db = connectDb(args);

  flexiblePositional(args, "field", "field", args.fieldless);

  const payloadData = handleDataArgs(args);

  const payloadWithState = await compileState(db, payloadData);

  const payload = addIdAndMetadata(payloadWithState, args);

  // update now in case the addDoc fails due to conflict
  await updateLastDocsRef(db, payload._id);

  const { undo, forceUndo } = args;
  if (undo || forceUndo) {
    return await primitiveUndo({
      db,
      payload,
      force: forceUndo,
      outputArgs: args,
    });
  }

  const conflictStrategy = args.conflict ?? (args.merge ? "merge" : undefined);
  const doc = await addDoc({
    db,
    payload,
    conflictStrategy,
    outputArgs: args,
  });

  // if addDoc has changed the id (e.g. because it relies on the modifiyTime), update lastDocRef again
  // TODO: if changing lastDoc to history may need to change this to overwrite first update
  if (doc._id !== payload._id) {
    await updateLastDocsRef(db, doc._id);
  }

  return doc;
}
