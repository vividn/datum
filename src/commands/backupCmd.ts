import { MainDatumArgs } from "../input/mainArgs";
import { connectDb } from "../auth/connectDb";
import * as fs from "fs";
import { DateTime } from "luxon";
import { dbArgs } from "../input/dbArgs";
import { ArgumentParser } from "argparse";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const backupArgs = new ArgumentParser({
  add_help: false,
});
// TODO: make this optional and then use stdout if not specified
backupArgs.add_argument("filename", {
  help: "file to write the backup to",
  type: "str",
});
backupArgs.add_argument("--overwrite", {
  help: "overwrite file if it exists",
  action: "store_true",
});

export const backupCmdArgs = new ArgumentParser({
  description: "Backup db, outputs to stdout",
  prog: "datum backup",
  usage: "%(prog)s <filename>",
  parents: [backupArgs, dbArgs],
});

export type BackupCmdArgs = MainDatumArgs & {
  filename: string;
  overwrite?: boolean;
};

export async function backupCmd(
  args: BackupCmdArgs | string | string[],
  preparsed?: Partial<BackupCmdArgs>,
): Promise<void> {
  args = parseIfNeeded(backupCmdArgs, args, preparsed);
  if (!args.overwrite && fs.existsSync(args.filename)) {
    throw new Error("File exists, overwrite with --overwrite");
  }
  const db = connectDb(args);
  const allDocs = (await db.allDocs({ include_docs: true })).rows.map(
    ({ doc }) => doc,
  );
  // TODO: Also backup and restore attachments (even though using attachments is considered not best practice)
  const backupTime = DateTime.utc().toISO() as string;
  const backupData = JSON.stringify({ backupTime, docs: allDocs }, null, 0);
  fs.writeFileSync(args.filename, backupData);
}
