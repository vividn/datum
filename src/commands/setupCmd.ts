import { Argv } from "yargs";
import { BaseDatumArgs } from "../input/baseYargs";
import { setupDatumViews } from "../views/setupDatumViews";
import { connectDb } from "../auth/connectDb";
import { Show } from "../output/output";

export const command = "setup";
export const desc = "setup the database for use with datum";

export type SetupCmdArgs = BaseDatumArgs & {
  projectDir?: string;
};

export function builder(yargs: Argv): Argv {
  return yargs.options({
    "project-dir": {
      describe:
        "where to look for additional DatumViews to setup in the database",
      type: "string",
    },
  });
}

export async function setupCmd(args: SetupCmdArgs): Promise<void> {
  const db = connectDb(args);
  const show: Show = args.showAll ? Show.All : args.show ?? Show.None;
  await setupDatumViews({ db, show: show, projectDir: args.projectDir });
}
