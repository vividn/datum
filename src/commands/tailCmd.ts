import { Argv } from "yargs";
import { EitherDocument } from "../documentControl/DatumDocument";
import { viewMap } from "../views/viewMap";
import { connectDb } from "../auth/connectDb";
import { interpolateFields } from "../utils/interpolateFields";
import { MainDatumArgs } from "../input/mainYargs";
import { FieldArgs, fieldArgs } from "../input/fieldArgs";
import { pullOutData } from "../utils/pullOutData";
import { extractFormatted } from "../output/output";
import Table from "easy-table";
import {
  TIME_METRICS,
  timingView,
  TimingViewType,
} from "../views/datumViews/timingView";
import { HIGH_STRING } from "../utils/startsWith";
import { handleTimeArgs, TimeArgs, timeYargs } from "../input/timeArgs";
import { reverseViewParams } from "../utils/reverseViewParams";
import { Show } from "../input/outputArgs";
import { DateTime } from "luxon";

export const command = ["tail [field]"];
export const desc =
  "show the most recently occurred/modified/created entries in the db";

export type TailCmdArgs = MainDatumArgs &
  TimeArgs &
  FieldArgs & {
    n?: number;
    metric?: "hybrid" | "occur" | "create" | "modify";
    head?: boolean;
  };

export function tailCmdYargs(yargs: Argv): Argv {
  return timeYargs(fieldArgs(yargs)).options({
    n: {
      alias: ["number"],
      describe: "number of entries to show, defaults to 10",
      type: "number",
    },
    metric: {
      describe:
        "which time to use for the sorting, default is hybrid: occur or modify",
      choices: TIME_METRICS,
      alias: "m",
      type: "string",
    },
    head: {
      describe: "show first rows instead of last rows",
      type: "boolean",
      hidden: true,
    },
  });
}

export const builder = tailCmdYargs;

export async function tailCmd(args: TailCmdArgs): Promise<EitherDocument[]> {
  const db = connectDb(args);

  const limit = args.n ?? 10;
  const metric = args.metric ?? "hybrid";
  const field = args.field ?? null;

  let viewParams: PouchDB.Query.Options<any, any> = {
    include_docs: true,
    inclusive_end: true,
    limit,
  };

  const { timeStr, unmodified: isDefaultTime, onlyDate } = handleTimeArgs(args);

  if (isDefaultTime || timeStr === undefined) {
    viewParams.startkey = [metric, field, ""];
    viewParams.endkey = [metric, field, HIGH_STRING];
  } else if (onlyDate) {
    // due to timezone shenanigans, must also grab the full days around the requested date and then filter later
    viewParams.limit = undefined;
    viewParams.startkey = [
      metric,
      field,
      DateTime.fromISO(timeStr)
        .minus({ day: 1 })
        .startOf("day")
        .toUTC()
        .toISO(),
    ];
    viewParams.endkey = [
      metric,
      field,
      DateTime.fromISO(timeStr).plus({ day: 1 }).endOf("day").toUTC().toISO(),
    ];
  } else if (args.head) {
    viewParams.startkey = [metric, field, timeStr];
    viewParams.endkey = [metric, field, HIGH_STRING];
  } else {
    viewParams.startkey = [metric, field, ""];
    viewParams.endkey = [metric, field, timeStr];
  }

  if (args.head !== true) {
    viewParams = reverseViewParams(viewParams);
  }
  const viewResults = await viewMap({
    db,
    datumView: timingView,
    params: viewParams,
  });

  // TODO factor this out better and automatically extract types from views
  const rawRows: {
    key: TimingViewType["MapKey"];
    value: TimingViewType["MapValue"];
    doc?: EitherDocument;
  }[] = args.head ? viewResults.rows : viewResults.rows.reverse();
  const filteredRows = onlyDate
    ? rawRows
        .filter((row) => row.value[0] === timeStr)
        // this sort moves times that are just dates to the top
        .sort((a, b) => (a.value >= b.value ? 1 : -1))
    : rawRows;
  const limitedRows =
    onlyDate && args.n === undefined
      ? filteredRows
      : args.head
      ? filteredRows.slice(0, limit)
      : filteredRows.slice(-limit);
  const docs: EitherDocument[] = limitedRows.map((row) => row.doc!);

  const format = args.formatString;
  const show = args.show;
  if (format && show !== Show.None) {
    docs.forEach((doc) => {
      const { data, meta } = pullOutData(doc);
      console.log(
        interpolateFields({ data, meta, format, useHumanTimes: true })
      );
    });
    return docs;
  }
  if (format === undefined && show !== Show.None) {
    const formattedRows = docs.map((doc) => {
      const formatted = extractFormatted(doc);
      return {
        time: formatted.time?.[metric],
        field: formatted.field,
        state: formatted.state,
        duration: formatted.dur,
        hid: formatted.hid,
      };
    });
    const headerRow = {
      time: "Time",
      duration: formattedRows.some((row) => row.duration !== undefined)
        ? "Dur"
        : undefined,
      field: "Field",
      state: formattedRows.some((row) => row.state !== undefined)
        ? "State"
        : undefined,
      hid: "HID",
    };

    const allRows = [headerRow, ...formattedRows];
    console.log(
      Table.print(allRows, { time: { printer: Table.padLeft } }, (table) => {
        return table.print();
      })
    );
  }
  return docs;
}
