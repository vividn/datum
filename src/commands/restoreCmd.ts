import { connectDb } from "../auth/connectDb";
import { readFileSync } from "fs";
import { dbArgs } from "../input/dbArgs";
import { ArgumentParser } from "argparse";
import { BackupCmdArgs } from "./backupCmd";
import { parseIfNeeded } from "../utils/parseIfNeeded";
import { MainDatumArgs } from "../input/mainArgs";

export const restoreArgs = new ArgumentParser({
  add_help: false,
});
restoreArgs.add_argument("filename", {
  help: "file to restore from",
  type: "str",
});
restoreArgs.add_argument("--allow-nonempty", {
  help: "Allow restore even if the db is not empty",
  action: "store_true",
  dest: "allowNonempty",
});

export const restoreCmdArgs = new ArgumentParser({
  description:
    "Restore db from backup file, for now only use on a new empty db for best results",
  prog: "datum restore",
  usage: "%(prog)s <filename>",
  parents: [restoreArgs, dbArgs],
});
export type RestoreCmdArgs = MainDatumArgs & {
  filename: string;
  allowNonempty?: boolean;
};
export async function restoreCmd(
  args: RestoreCmdArgs | string | string[],
  preparsed?: Partial<BackupCmdArgs>,
): Promise<void> {
  args = parseIfNeeded(restoreCmdArgs, args, preparsed);
  args.createDb ??= true;
  const db = connectDb(args);
  const info = await db.info();
  if (!args.allowNonempty && info.doc_count !== 0) {
    throw new Error(
      "Warning: db is not empty, aborting restore. Use --allow-nonempty to override.",
    );
  }
  const buffer = readFileSync(args.filename);
  const allDocs = JSON.parse(buffer.toString()).docs;
  await db.bulkDocs(allDocs, { new_edits: false });
}
