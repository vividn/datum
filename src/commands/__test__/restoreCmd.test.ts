import { restoreCmd } from "../restoreCmd";
import { testDbLifecycle } from "../../test-utils";
import path from "path";
import * as os from "os";
import * as fs from "fs";
import { generateSampleMorning } from "./tailCmd.test";
import { setupCmd } from "../setupCmd";
import { backupCmd } from "../backupCmd";
import * as connectDbModule from "../../auth/connectDb";
import { connectDb } from "../../auth/connectDb";
import { addCmd } from "../addCmd";

describe("restoreCmd", () => {
  const dbName1 = "restore_cmd_test_1";
  const dbName2 = "restore_cmd_test_2";
  const db2 = testDbLifecycle(dbName2);

  // this one takes precedence for the default db
  const db1 = testDbLifecycle(dbName1);

  let backupFilePath: string;
  beforeEach(async () => {
    const tempDir = process.env["RUNNER_TEMP"] ?? os.tmpdir();
    backupFilePath = path.join(tempDir, "backup.json");
    if (fs.existsSync(backupFilePath)) {
      fs.unlinkSync(backupFilePath);
    }
    await setupCmd({});
    await generateSampleMorning("2023-11-17");
    expect((await db1.info()).doc_count).toBeGreaterThan(5);
    expect((await db2.info()).doc_count).toBe(0);

    await backupCmd({ filename: backupFilePath });
  });
  afterEach(() => {
    if (fs.existsSync(backupFilePath)) {
      fs.unlinkSync(backupFilePath);
    }
  });

  it("can restore a previously backed up db to a new empty db, having exactly the same docs and revisions", async () => {
    expect((await db2.info()).doc_count).toBe(0);

    jest.spyOn(connectDbModule, "connectDb").mockReturnValueOnce(db2);
    await restoreCmd({
      filename: backupFilePath,
    });
    expect((await db2.info()).doc_count).not.toBe(0);

    const restoredDocs = (await db2.allDocs({ include_docs: true })).rows.map(
      ({ doc }) => doc,
    );
    const originalDocs = (await db1.allDocs({ include_docs: true })).rows.map(
      ({ doc }) => doc,
    );
    expect(restoredDocs.length).toBe(originalDocs.length);
    expect(restoredDocs).toEqual(originalDocs);
  });

  it("creates a new db if it does not exist", async () => {
    const unmadeDbName = "restore_cmd_test_unmade_db";
    const spy = jest.spyOn(connectDbModule, "connectDb");
    spy.mockRestore();
    const unmadeDb = connectDb({ db: unmadeDbName, createDb: false });
    await restoreCmd({
      db: unmadeDbName,
      filename: backupFilePath,
    });
    expect((await unmadeDb.info()).doc_count).not.toBe(0);
    await unmadeDb.destroy();
  });

  it("throws an error if the db is not empty", async () => {
    expect((await db2.info()).doc_count).toBe(0);
    await db2.put({ _id: "one_document" });

    jest.spyOn(connectDbModule, "connectDb").mockReturnValueOnce(db2);
    await expect(
      restoreCmd({
        filename: backupFilePath,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("can restore over a nonempty db with the appropriate option", async () => {
    expect((await db2.info()).doc_count).toBe(0);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db2);
    const extraDoc = await addCmd({ field: "extra_doc" });

    await restoreCmd({
      filename: backupFilePath,
      allowNonempty: true,
    });

    const restoredDocs = (await db2.allDocs({ include_docs: true })).rows.map(
      ({ doc }) => doc,
    );
    const originalDocs = (await db1.allDocs({ include_docs: true })).rows.map(
      ({ doc }) => doc,
    );
    expect(restoredDocs.length).toBe(originalDocs.length + 1);
    expect(restoredDocs).toContainEqual(extraDoc);
    expect(restoredDocs).toContainEqual(originalDocs.at(1));
  });
});
