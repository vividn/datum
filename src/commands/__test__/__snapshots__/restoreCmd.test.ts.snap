// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restoreCmd throws an error if the db is not empty 1`] = `"Warning: db is not empty, aborting restore. Use --allow-nonempty to override."`;
